src/main/scala/io/chrisdavenport/mules/AutoFetchingCache.scala

```diff
@@ -0,0 +1,195 @@
+    cache.values.get.map(_.keys.toList)
+
+  /**
+    * Delete an item from the cache. Won't do anything if the item is not present.
+    **/
+  def delete[F[_], K, V](cache: AutoFetchingCache[F, K, V])(k: K)
+                        (implicit S: Sync[F]): F[Unit] =
+    cache.values.modify(m =>
+      (m - k, m.get(k) match {
+        case Some(Fetching(f)) => f.cancel
+        case _ => S.unit
+      })) >> cache.reload.traverse { case (_, values) =>
+      values.modify(m => (m - k,
+        m.get(k) match {
+          case Some(f) => f.cancel
+          case None => S.unit
+        }
+      ))
+    }.void
+
+
+
+  def cancelReloads[F[_] : Monad, K, V](cache: AutoFetchingCache[F, K, V]): F[Unit] =
+    cache.reload.map { case (_, values) =>
+      values.modify(m => (Map.empty,
+        m.values.toList.traverse[F, Unit](_.cancel).void)
+      ).flatten
+    } getOrElse Monad[F].unit
+
+  /**
+    * Insert an item in the cache, using the default expiration value of the cache.
+    */
+  private def insert[F[_] : Sync : Timer, K, V](cache: AutoFetchingCache[F, K, V])(k: K, v: V): F[Unit] =
+    insertWithTimeout(cache)(cache.defaultExpiration)(k, v)
+
+
+  /**
+    * Insert an item in the cache, with an explicit expiration value.
+    *
+    * If the expiration value is None, the item will never expire. The default expiration value of the cache is ignored.
+    *
+    * The expiration value is relative to the current clockMonotonic time, i.e. it will be automatically added to the result of clockMonotonic for the supplied unit.
+    **/
+
+  def insertWithTimeout[F[_] : Sync, K, V](cache: AutoFetchingCache[F, K, V])
+                                          (optionTimeout: Option[TimeSpec])
+                                          (k: K, v: V)
+                                          (implicit T: Timer[F]): F[Unit] = {
+    for {
+      now <- T.clock.monotonic(NANOSECONDS)
+      timeout = optionTimeout.map(ts => TimeSpec.unsafeFromNanos(now + ts.nanos))
+      _ <- cache.values.update(_ + (k -> CacheItem[F, V](v, timeout)))
+    } yield ()
+  }
+
+
+  private def insertFetching[F[_], K, V](k: K, cache: AutoFetchingCache[F, K, V])
+                                        (f: Fiber[F, V]): F[Unit] = {
+    cache.values.update(_ + (k -> Fetching[F, V](f)))
+  }
+
+
+  private def isExpired[F[_], A](checkAgainst: TimeSpec, cacheItem: CacheItem[F, A]): Boolean = {
+    cacheItem.itemExpiration.fold(false) {
+      case e if e.nanos < checkAgainst.nanos => true
+      case _ => false
+    }
+  }
+
+  private def lookupItemSimple[F[_] : Sync, K, V](k: K, c: AutoFetchingCache[F, K, V]): F[Option[CacheContent[F, V]]] =
+    c.values.get.map(_.get(k))
+
+  private def fetchAndInsert[F[_] : Sync : Timer, K, V](k: K, c: AutoFetchingCache[F, K, V]): F[V] =
+    for {
+      v <- c.fetch(k)
+      _ <- insert(c)(k, v)
+    } yield v
+
+
+  private def autoReload[F[_], K, V](k: K, c: AutoFetchingCache[F, K, V])
```

Comment:
What semantics are you trying to provide via this function?

Reply:
this function setup what I understand being a [refresh](https://github.com/google/guava/wiki/CachesExplained#refresh) behavior. It will fetch a newer value every reload period. After writing it I realized I must add the capacity to configure an upper bound of the number of reload that can be active concurrently. I'll try to add it asap but it will change this part of the code quiet a lot I think.
I have a problem that forces me to check if a previous value is already set : I cannot check the existence of the key and modify the map atomically. Hence if a given key is fetched several times on the time of its first creation, several check can return true.